File: ticket_tool/poc_app1/test/jira/go/main.go
package main

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
)

type Config struct {
	JiraBaseUrl  string `json:"jira_base_url"`
	JiraUsername string `json:"jira_username"`
	JiraApiToken string `json:"jira_api_token"`
}

type Project struct {
	ID             string `json:"id"`
	Key            string `json:"key"`
	Name           string `json:"name"`
	ProjectTypeKey string `json:"projectTypeKey"`
}

type ProjectResponse struct {
	Values     []Project `json:"values"`
	StartAt    int       `json:"startAt"`
	MaxResults int       `json:"maxResults"`
	Total      int       `json:"total"`
}

func getBasicAuthHeader(username, apiToken string) string {
	auth := username + ":" + apiToken
	return "Basic " + base64.StdEncoding.EncodeToString([]byte(auth))
}

func getProjects(config Config) (*ProjectResponse, error) {
	client := &http.Client{}
	url := config.JiraBaseUrl + "/rest/api/3/project/search"

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("error creating request: %v", err)
	}

	req.Header.Add("Authorization", getBasicAuthHeader(config.JiraUsername, config.JiraApiToken))
	req.Header.Add("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("error making request: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("error response from JIRA: status=%d, body=%s", resp.StatusCode, string(body))
	}

	var projectResp ProjectResponse
	if err := json.NewDecoder(resp.Body).Decode(&projectResp); err != nil {
		return nil, fmt.Errorf("error decoding response: %v", err)
	}

	return &projectResp, nil
}

func handleProjects(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	file, err := os.Open("config.json")
	if err != nil {
		http.Error(w, "Error opening config file: "+err.Error(), http.StatusInternalServerError)
		return
	}
	defer file.Close()

	var config Config
	if err := json.NewDecoder(file).Decode(&config); err != nil {
		http.Error(w, "Error decoding config file: "+err.Error(), http.StatusInternalServerError)
		return
	}

	projects, err := getProjects(config)
	if err != nil {
		http.Error(w, "Error getting projects: "+err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(projects)
}

func main() {
	http.HandleFunc("/api/projects", handleProjects)

	port := "8080"
	fmt.Printf("Server starting on port %s...\n", port)
	if err := http.ListenAndServe(":"+port, nil); err != nil {
		log.Fatal(err)
	}
}

File: ticket_tool/poc_app1/test/jira/rust/src/main.rs
use base64::{engine::general_purpose::STANDARD, Engine};
use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION, CONTENT_TYPE};
use serde::{Deserialize, Serialize};
use std::error::Error;
use tokio::fs;

#[derive(Debug, Deserialize)]
struct Config {
    api: ApiConfig,
}

#[derive(Debug, Deserialize)]
struct ApiConfig {
    jira_base_url: String,
    jira_username: String,
    jira_api_token: String,
}

#[derive(Debug, Deserialize, Serialize)]
struct Project {
    id: String,
    key: String,
    name: String,
    #[serde(rename = "projectTypeKey")]
    project_type_key: String,
}

async fn request(url: &str, headers: HeaderMap) -> Result<String, Box<dyn Error>> {
    let client = reqwest::Client::new();
    let response = client
        .get(url)
        .headers(headers)
        .send()
        .await?;

    println!("Status: {}", response.status());
    
    if !response.status().is_success() {
        return Err(format!("Request failed with status: {}", response.status()).into());
    }

    let body = response.text().await?;
    Ok(body)
}

async fn create_headers(username: &str, api_token: &str) -> Result<HeaderMap, Box<dyn Error>> {
    let mut headers = HeaderMap::new();
    
    // Basic認証ヘッダーの作成
    let auth = format!("{}:{}", username, api_token);
    let encoded_auth = STANDARD.encode(auth.as_bytes());
    let auth_header = format!("Basic {}", encoded_auth);
    headers.insert(
        AUTHORIZATION,
        HeaderValue::from_str(&auth_header)?
    );

    // Content-Typeヘッダーの設定
    headers.insert(
        CONTENT_TYPE,
        HeaderValue::from_static("application/json")
    );

    // Accept ヘッダーの追加
    headers.insert(
        reqwest::header::ACCEPT,
        HeaderValue::from_static("application/json")
    );
    Ok(headers)
}

async fn request_api(headers: HeaderMap, url: &str) {
    println!("Requesting projects from: {}", url);
    let response = request(url, headers).await;
    match response {
        Ok(_body) => {
            // println!("Response: {}", body);
        },
        Err(e) => {
            println!("Error: {}", e);
        }
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // 設定ファイルを読み込む
    let config_data = fs::read_to_string("config.toml").await?;
    let config: Config = toml::from_str(&config_data)?;
    
    // ヘッダーを作成
    let headers = create_headers(&config.api.jira_username, &config.api.jira_api_token).await?;

    // JIRAのプロジェクト一覧を取得
    let projects_url = format!("{}/rest/api/3/project", config.api.jira_base_url);
    request_api(headers, &projects_url).await;


    Ok(())
}
